File: app/api/routes/registry.py
from __future__ import annotations

from fastapi import APIRouter, Depends, File, HTTPException, Query, UploadFile
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import CurrentUser, get_db
from app.db.repositories import RegistryRepository
from app.schemas.registry import RegistryEntryListRead, RegistryEntryRead
from app.services.registry_ingest import (
    REGISTRY_SERIAL_KEYS,
    ingest_registry_rows,
)
from app.utils.excel import read_rows_with_required_headers

router = APIRouter(prefix="/api/v1/registry", tags=["registry"])


@router.post("/import")
async def import_registry_file(
    db_file: UploadFile = File(...),
    *,
    source_sheet: str | None = Query(
        None,
        description="Имя листа в Excel, если нужно жёстко указать",
    ),
    instrument_kind: str | None = Query(
        None,
        description="Тип средств измерений (например, controllers, manometers)",
    ),
    header_row: int = Query(5, ge=1, description="Номер строки с заголовками"),
    data_start_row: int = Query(6, ge=1, description="Первая строка с данными"),
    session: AsyncSession = Depends(get_db),
):
    payload = await db_file.read()
    if not payload:
        raise HTTPException(status_code=400, detail="empty registry file")

    rows = read_rows_with_required_headers(
        payload,
        header_row=header_row,
        data_start_row=data_start_row,
        required_headers=REGISTRY_SERIAL_KEYS,
    )
    if not rows:
        raise HTTPException(status_code=400, detail="registry file has no serial entries")

    result = await ingest_registry_rows(
        session,
        source_file=db_file.filename or "registry.xlsx",
        rows=rows,
        source_sheet=source_sheet,
        instrument_kind=instrument_kind,
    )

    await session.commit()

    return {
        "processed": result["processed"],
        "deactivated": result["deactivated"],
        "instrument_kind": instrument_kind or source_sheet,
        "source_file": db_file.filename or "registry.xlsx",
    }


@router.get("/entries", response_model=RegistryEntryListRead)
async def list_registry_entries(
    _: CurrentUser,
    *,
    search: str | None = Query(
        None,
        min_length=1,
        description="Свободный поиск по серийнику, документу или источнику",
    ),
    instrument_kind: str | None = Query(None, description="Фильтр по типу прибора"),
    active_only: bool = Query(True, description="Показывать только активные записи"),
    limit: int = Query(300, ge=1, le=1000),
    session: AsyncSession = Depends(get_db),
) -> RegistryEntryListRead:
    total, entries = await RegistryRepository(session).list_entries(
        search=search,
        instrument_kind=instrument_kind,
        active_only=active_only,
        limit=limit,
    )
    return RegistryEntryListRead(
        total=total,
        items=[
            RegistryEntryRead(
                id=entry.id,
                source_file=entry.source_file,
                source_sheet=entry.source_sheet,
                instrument_kind=entry.instrument_kind,
                row_index=entry.row_index,
                normalized_serial=entry.normalized_serial,
                document_no=entry.document_no,
                protocol_no=entry.protocol_no,
                owner_name_raw=entry.owner_name_raw,
                methodology_raw=entry.methodology_raw,
                verification_date=entry.verification_date,
                valid_to=entry.valid_to,
                is_active=entry.is_active,
                loaded_at=entry.loaded_at,
            )
            for entry in entries
        ],
    )

File: frontend/src/api/registry.ts
import { apiRequest } from "@/api/client";

type RawRegistryImportResponse = {
  processed: number;
  deactivated: number;
  instrument_kind: string | null;
  source_file: string;
};

type RawRegistryEntry = {
  id: number;
  source_file: string;
  source_sheet?: string | null;
  instrument_kind?: string | null;
  row_index: number;
  normalized_serial?: string | null;
  document_no?: string | null;
  protocol_no?: string | null;
  owner_name_raw?: string | null;
  methodology_raw?: string | null;
  verification_date?: string | null;
  valid_to?: string | null;
  is_active: boolean;
  loaded_at: string;
};

type RawRegistryEntryList = {
  total: number;
  items: RawRegistryEntry[];
};

export type RegistryImportPayload = {
  file: File;
  sourceSheet?: string;
  instrumentKind?: string;
  headerRow?: number;
  dataStartRow?: number;
};

export type RegistryImportResponse = {
  processed: number;
  deactivated: number;
  instrumentKind: string | null;
  sourceFile: string;
};

export type RegistryEntry = {
  id: number;
  sourceFile: string;
  sourceSheet: string | null;
  instrumentKind: string | null;
  rowIndex: number;
  serial: string | null;
  documentNo: string | null;
  protocolNo: string | null;
  ownerName: string | null;
  methodology: string | null;
  verificationDate: string | null;
  validTo: string | null;
  isActive: boolean;
  loadedAt: string;
};

export type RegistryEntryList = {
  total: number;
  items: RegistryEntry[];
};

export async function importRegistryFile(
  token: string,
  payload: RegistryImportPayload,
): Promise<RegistryImportResponse> {
  const search = new URLSearchParams();
  if (payload.sourceSheet?.trim()) {
    search.set("source_sheet", payload.sourceSheet.trim());
  }
  if (payload.instrumentKind?.trim()) {
    search.set("instrument_kind", payload.instrumentKind.trim());
  }
  if (payload.headerRow) {
    search.set("header_row", String(payload.headerRow));
  }
  if (payload.dataStartRow) {
    search.set("data_start_row", String(payload.dataStartRow));
  }

  const formData = new FormData();
  formData.append("db_file", payload.file);

  const response = await apiRequest<RawRegistryImportResponse>(
    `/registry/import${search.size ? `?${search.toString()}` : ""}`,
    {
      method: "POST",
      token,
      body: formData,
    },
  );

  return {
    processed: response.processed,
    deactivated: response.deactivated,
    instrumentKind: response.instrument_kind,
    sourceFile: response.source_file,
  };
}

export async function fetchRegistryEntries(
  token: string,
  params: {
    search?: string;
    instrumentKind?: string;
    activeOnly?: boolean;
    limit?: number;
  } = {},
): Promise<RegistryEntryList> {
  const search = new URLSearchParams();
  if (params.search?.trim()) {
    search.set("search", params.search.trim());
  }
  if (params.instrumentKind?.trim()) {
    search.set("instrument_kind", params.instrumentKind.trim());
  }
  if (params.activeOnly !== undefined) {
    search.set("active_only", String(params.activeOnly));
  }
  if (params.limit) {
    search.set("limit", String(params.limit));
  }

  const response = await apiRequest<RawRegistryEntryList>(
    `/registry/entries${search.size ? `?${search.toString()}` : ""}`,
    {
      method: "GET",
      token,
    },
  );

  return {
    total: response.total,
    items: response.items.map((item) => ({
      id: item.id,
      sourceFile: item.source_file,
      sourceSheet: item.source_sheet ?? null,
      instrumentKind: item.instrument_kind ?? null,
      rowIndex: item.row_index,
      serial: item.normalized_serial ?? null,
      documentNo: item.document_no ?? null,
      protocolNo: item.protocol_no ?? null,
      ownerName: item.owner_name_raw ?? null,
      methodology: item.methodology_raw ?? null,
      verificationDate: item.verification_date ?? null,
      validTo: item.valid_to ?? null,
      isActive: item.is_active,
      loadedAt: item.loaded_at,
    })),
  };
}
